### PR TITLE
feat(babel): react pure components

### PR DESCRIPTION
**Pull request by justingreenberg**
[babel-plugin-react-pure-components](https://github.com/thejameskyle/babel-plugin-react-pure-components) - this is essentially a workaround until there's a stable solution for instrumenting pure function components. 

this plugin transforms pure component classes into function components in production builds—that is, react classes with only a `render` method (no lifecycle methods such as `shouldComponentUpdate`). the plugin currently supports static `defaultProps` and `propTypes` as well. 

reasoning:
 - allows use of pure component classes with full transform and tooling support in development
 - transform pure component classes into to pure functions for optimized production builds.

example:
```js
class HomeView extends React.Component {
  static propTypes = {
    counter: PropTypes.number.isRequired,
    doubleAsync: PropTypes.func.isRequired,
    increment: PropTypes.func.isRequired
  };

  render () {
    return (
      <div className='container text-center'>
        <div className='row'>
          <div className='col-xs-2 col-xs-offset-5'>
            <img className={classes.duck}
                 src={DuckImage}
                 alt='This is a duck, because Redux.' />
          </div>
        </div>
        <h1>Welcome to the React Redux Starter Kit</h1>
        <h2>
          Sample Counter:&nbsp;
          <span className={classes['counter--green']}>{this.props.counter}</span>
        </h2>
        <button className='btn btn-default'
                onClick={() => this.props.increment(1)}>
          Increment
        </button>
        <button className='btn btn-default'
                onClick={this.props.doubleAsync}>
          Double (Async)
        </button>
        <hr />
        <Link to='/404'>Go to 404 Page</Link>
      </div>
    )
  }
}
``` 
**production** output:
```js
function HomeView(props) {
  return React.createElement(
    'div',
    { className: 'container text-center' },
    React.createElement(
      'div',
      { className: 'row' },
      React.createElement(
        'div',
        { className: 'col-xs-2 col-xs-offset-5' },
        React.createElement('img', { className: classes.duck,
          src: DuckImage,
          alt: 'This is a duck, because Redux.' })
      )
    ),
    React.createElement(
      'h1',
      null,
      'Welcome to the React Redux Starter Kit'
    ),
    React.createElement(
      'h2',
      null,
      'Sample Counter: ',
      React.createElement(
        'span',
        { className: classes['counter--green'] },
        props.counter
      )
    ),
    React.createElement(
      'button',
      { className: 'btn btn-default',
        onClick: function onClick() {
          return props.increment(1);
        } },
      'Increment'
    ),
    React.createElement(
      'button',
      { className: 'btn btn-default',
        onClick: props.doubleAsync },
      'Double (Async)'
    ),
    React.createElement('hr', null),
    React.createElement(
      Link,
      { to: '/404' },
      'Go to 404 Page'
    )
  );
}

HomeView.propTypes = {
  counter: PropTypes.number.isRequired,
  doubleAsync: PropTypes.func.isRequired,
  increment: PropTypes.func.isRequired
};
```

dan abramov has hinted that he is working on a new method for hot reloading that would resolve this issue. at that point we can lose the classes altogether and leverage toolkits based on functional composition—like recompose (or recompose-rx)—which I think will ultimately replace react-redux as the preferred method of binding to redux stores.

plugin was written by one of the authors of babel

### CODE REVIEW

File: src/views/HomeView/HomeView.js [hunk 1]
```diff
@@ -13,7 +13,7 @@ import classes from './HomeView.scss'
 const mapStateToProps = (state) => ({
   counter: state.counter
 })
-export class HomeView extends React.Component {
+class HomeView extends React.Component {
```
Comment (justingreenberg) [inline src/views/HomeView/HomeView.js]: tracking thejameskyle/babel-plugin-react-pure-components#5

export syntax (on line 54 below) is slightly modified awaiting hotfix above—will be changed back to a named es module export (`export class HomeView extends React.Component {}`) as soon as pr above is merged

**Comment by davezuko** [comment]
Wow that's a super cool plugin, but a bit of a shame that it's needed (though understandable). Do you think we should hold off on this until the referenced PR is merged to fix the named exports issue? I'm a bit hesitant to introduce a strange workflow just to accommodate a soon-to-be-fixed bug.

That aside, :+1: for getting this in.

**Comment by justingreenberg** [comment]
i agree, i just followed up on PR so let's see what james says. it looks like the issue is coming from parse, and the maintainer is core babel contributor, so he may be trying to research the root cause as opposed to merging hotfix

**Comment by davezuko** [comment]
Just saw that, thank you!

Need to get you a :trophy:

**Comment by davezuko** [comment]
Still awaiting that PR mergeroni.

**Comment by justingreenberg** [comment]
@davezuko hope all is well! closing for time being, export bugs are still in babel core...

i know we've both been busy with work, but i have some cool ideas to bounce off you, i'll try to open some issues for discussion later this week :)

**Comment by davezuko** [comment]
@justingreenberg that's a shame, would love to pick it back up once things stabilize.

Haven't heard from you in a while, so I hope things are going well with you as well. Wish the FE world had more than gitter and ticket systems, haha. Anywho, looking forward to your ideas, thanks again!

**Pull request closed without merging**